docs/source/dev_guide/algorithms/implicit_als/computation-distributed-training.rst

```diff
@@ -0,0 +1,246 @@
+
+.. _implicit_als_distributed_training_step_2:
+
+Step 2 - on Master Node
+***********************
+
+This step uses local partial results from :ref:`Step 1 <implicit_als_distributed_training_step_1>` as input.
+
+.. image:: images/implicit-als-distributed-computation-training-step-2.png
+    :width: 600
+    :align: center
+
+In this step, implicit ALS recommender training accepts the input described below.
+Pass the ``Input ID`` as a parameter to the methods that provide input for your algorithm.
+For more details, see :ref:`algorithms`.
+
+.. list-table::
+   :widths: 10 60
+   :header-rows: 1
+
+   * - Input ID
+     - Input
+   * - ``inputOfStep2FromStep1``
+     - A collection of numeric tables computed on local nodes in :ref:`Step 1 <implicit_als_distributed_training_step_1>`.
+       
+       .. note::
+            The collection may contain objects of any class derived from ``NumericTable``
+            except the ``PackedTriangularMatrix`` class with the ``lowerPackedTriangularMatrix`` layout.
+
+In this step, implicit ALS recommender training calculates the result described below.
+Pass the ``Result ID`` as a parameter to the methods that access the results of your algorithm.
+For more details, see :ref:`algorithms`.
+
+.. list-table::
+   :widths: 10 60
+   :header-rows: 1
+   :align: left
+
+   * - Result ID
+     - Result
+   * - ``outputOfStep2ForStep4``
+     - Pointer to the :math:`f \times f` numeric table with merged cross-products.
+
+.. _implicit_als_distributed_training_step_3:
+
+Step 3 - on Local Nodes
+***********************
+
+On each node :math:`i`, this step uses results of the previous steps and requires
+that you provide two extra matrices Offset Table i and Input of Step 3 From Init i
+computed at the initialization stage of the algorithm.
+
+The only element of the Offset Table i table refers to the:
+
+- :math:`i`-th element of the ``offsets`` collection from the
+  :ref:`step 2 of the distributed initialization algorithm <implicit_als_distributed_init_step_2>` in :ref:`part 1 <implicit_als_computation_parts>` of the iteration
+- :math:`i`-th element of the ``offsets`` collection from the
+  :ref:`step 1 of the distributed initialization algorithm <implicit_als_distributed_init_step_1>` in :ref:`part 2 <implicit_als_computation_parts>` of the iteration
+
+The Input Of Step 3 From Init is a key-value data collection that refers to the ``outputOfInitForComputeStep3`` output of the initialization stage:
+
+- Output of the :ref:`step 1 of the distributed initialization algorithm <implicit_als_distributed_init_step_1>` in :ref:`part 1 <implicit_als_computation_parts>` of the iteration
+- Output of the :ref:`step 2 of the distributed initialization algorithm <implicit_als_distributed_init_step_2>` in :ref:`part 2 <implicit_als_computation_parts>` of the iteration
+
+.. image:: images/implicit-als-distributed-computation-training-step-3.png
+    :width: 600
+    :align: center
+
```

Comment:
The text "In this step, implicit ALS recommender training accepts the input described below. ... " is missing here.

Reply:
fixed, thanks